zebra-consensus/src/checkpoint.rs

```diff
@@ -919,8 +919,7 @@ where
             match state_service
                 .oneshot(zs::Request::CommitFinalizedBlock(block.into()))
                 .map_err(VerifyCheckpointError::CommitFinalized)
-                .await
-                .expect("state service commit block failed: verified checkpoints must be committed transactionally")
+                .await?
             {
```

Comment:
When there's an error, should we reset the `verifier_progress` to the state finalized tip?

### Checkpoint Verifier & State

Zebra assumes that the state service and checkpoint verifier are at the same height.
But if there is a `CommitFinalized` error, then the checkpoint tip is later than the state tip.

So Zebra is unable to make any more sync progress, because:
1. the syncer downloads the block after `StateService.finalized_tip`
2. but `CheckpointVerifier.verifier_progress` rejects that block, because it's behind the checkpoint tip

One way to solve that issue is to panic.
But we don't want peers to be able to trigger panics by sending us bad blocks.

What do you think we should do here?

### Tip Reset Code

Here's the code that sets the initial tip:
https://github.com/ZcashFoundation/zebra/blob/eadca72e757ae074a91a3ff6cbed4ffd275e6849/zebra-consensus/src/checkpoint.rs#L215-L227

Here's `tokio`'s advice about communicating from async to sync code:
https://docs.rs/tokio/1.10.0/tokio/sync/mpsc/index.html#communicating-between-sync-and-async-code

Reply:
I agree that there is a problem with the current approach. But I'm kinda stuck with how to fix this.

What I understand:

When the block is comitted here:

https://github.com/ZcashFoundation/zebra/blob/5c5abf61719b73af46d54473741c482b585b48fc/zebra-consensus/src/checkpoint.rs#L919-L923

All the blocks in the range were already cleared to be committed in

https://github.com/ZcashFoundation/zebra/blob/5c5abf61719b73af46d54473741c482b585b48fc/zebra-consensus/src/checkpoint.rs#L731-L734

And it doesn't seem possible to stop that from going forward.

So we can:

- Do nothing; the other commits will fail (because their heights don't match the finalized state tip). But we would need to reset the `verifier_progress` somehow. And I still don't understand if just that is enough (will `ChainSync` redownload those blocks automatically? I don't fully understand that part yet)
- Somehow redesign how `ChainVerifier` handles these, by only removing the queued blocks and updating the progress after the blocks are properly committed. Not yet sure how.
- Change `ChainVerifier` to always query the state and having a single source of truth of where it stands, but this ends up being the option before (redesign)
- Give up on doing this in the finalized state and add trees in the checkpoint verifier (I think in the end this may be end up being simpler? But I'm still not 100% sure where to do this, see question below)
- Something else that I haven't considered.

And here's a random-ish question. How do we know that these channels will be received in the same sequence as they are sent? It seems the code relies on that being true.

https://github.com/ZcashFoundation/zebra/blob/5c5abf61719b73af46d54473741c482b585b48fc/zebra-consensus/src/checkpoint.rs#L731-L734

Reply:
>     * Do nothing; the other commits will fail (because their heights don't match the finalized state tip). But we would need to reset the `verifier_progress` somehow.

What about asking the state for the finalized tip, and setting `verifier_progress` to that tip?

> And I still don't understand if just that is enough (will `ChainSync` redownload those blocks automatically? I don't fully understand that part yet)

Yes. If there are any errors, `ChainSync` restarts syncing from the finalized tip.

> And here's a random-ish question. How do we know that these channels will be received in the same sequence as they are sent? It seems the code relies on that being true.

They're not received in order, so the finalized state queues them and commits them in order:
https://github.com/ZcashFoundation/zebra/blob/5c5abf61719b73af46d54473741c482b585b48fc/zebra-state/src/service/finalized_state.rs#L142-L146

Did you want to jump on a call some time and talk through the possible designs here?

Reply:
I implemented the first option in 9b7b0a15
It was relatively simple, but I don't know how to test it. Do you have any ideas?

Reply:
I have an idea, I'm finishing the test with fake activation heights and will describe it after I finish that

Reply:
I added the fake activation height test in a3ff5b26. It must be run like

```
cargo test --package zebra-state --lib --features test_fake_activation_heights -- service::finalized_state::tests::prop::all_upgrades_and_wrong_commitments --exact
```

It tests that the finalized state check is working, but it does not test the checkpoint verifier itself. (It could be used to test the non-finalized state too)

To test that, my idea is to basically:

- Create a new test based on [`continuous_blockchain`](https://github.com/ZcashFoundation/zebra/blob/79d41b3aea1081946f06095eb1f8d9d84cbb4c28/zebra-consensus/src/checkpoint/tests.rs#L232)
- Change it to corrupt some block and check if it verifier is correctly reset

So I'd like feedback on these:

- Does the fake activation height test that I added makes sense?
- How to test that on CI? We'd need to activate the feature and only run the tests that use it. The only way to filter that seems to be with some prefix or suffix in the test name itself?
- Does the checkpoint verifier test that I described makes sense?

Reply:
> * Does the fake activation height test that I added makes sense?

Looks good!

> * How to test that on CI? We'd need to activate the feature and only run the tests that use it. The only way to filter that seems to be with some prefix or suffix in the test name itself?

I think in CI we just run cargo test with the feature and the full test name.
(For example, see the `test.yml` workflow for the cached state tests.)

We might also want to check that at least one test runs, in case we change the test name.

> * Does the checkpoint verifier test that I described makes sense?

Yes, that sounds good too.

Reply:
> What is missing: ensuring the fake activation height tests are actually run, as you've mentioned. (Any suggestions on how to do this?)

Can we open a follow-up ticket for the general issue of making sure that selected tests run?
Because it's not just this test we'll need to check, it's the cached state tests as well.

(We can check the test output using grep, but maybe there's a better solution.)

> At this point it seems much more work than expected and maybe no worth it.

I'd like to merge this PR with the current set of working tests, and open a ticket in sprint 21 to test checkpoint rollback.

Here's the tradeoffs I'm seeing:
* If we don't check the authorizing data hash for checkpointed blocks, peers can send us bad blocks.
* If we do check, but don't completely test, Zebra might panic or get stuck rather than rolling back.
* If we spend a lot of time trying to get the tests right, we might not finish the mempool work on time.

Getting stuck and sending lots of messages to peers is my biggest concern, so that's why I'm suggesting tests in sprint 21. Even if we just test with block test vectors 0-10 and a fake checkpoint list.